File: backend/users/instruments.py
"""
The brass instruments a player can choose during onboarding.

A pure module (constants + lookups, no Django imports) in the same spirit as
``progress.rewards``, so the grading engine can import it later without touching
the database. Mirrored on the client by
``apps/mobile/src/data/instruments.ts`` — change both together.

Transposition convention matches ``studies.StudyContent.transposition_semitones``
(``-2`` for the B♭ trumpet parts the Clarke corpus is written in)::

    sounding pitch = written pitch + sounding_offset_semitones

Onboarding itself only reads ``slug``/``label``/``family``. ``clef`` and
``sounding_offset_semitones`` are recorded now as groundwork for the per-instrument
transposition work: they describe the instrument, not the app's use of it, so
writing them down once here keeps that data in a single reviewed place.
"""
from __future__ import annotations

from dataclasses import dataclass

# Notation the player reads from. The Clarke corpus is engraved in treble clef;
# the low brass entries below read the same music in bass clef.
CLEF_TREBLE = "treble"
CLEF_BASS = "bass"

# Coarse grouping, used to order and section the onboarding picker.
FAMILY_TRUMPET = "trumpet"
FAMILY_HORN = "horn"
FAMILY_LOW_BRASS = "low-brass"


@dataclass(frozen=True)
class Instrument:
    """One selectable instrument and the facts needed to transpose for it."""

    slug: str
    label: str
    clef: str
    # Semitones from written to sounding pitch; negative sounds below the page.
    sounding_offset_semitones: int
    family: str


# Ordered as the onboarding picker presents them: trumpet family first (the
# instrument Clarke wrote for), then horns, then low brass.
INSTRUMENTS: tuple[Instrument, ...] = (
    Instrument("trumpet", "Trumpet", CLEF_TREBLE, -2, FAMILY_TRUMPET),
    Instrument("cornet", "Cornet", CLEF_TREBLE, -2, FAMILY_TRUMPET),
    Instrument("flugelhorn", "Flugelhorn", CLEF_TREBLE, -2, FAMILY_TRUMPET),
    # B♭ piccolo trumpet sounds an octave above the B♭ trumpet: -2 + 12.
    Instrument("piccolo-trumpet", "Piccolo Trumpet", CLEF_TREBLE, 10, FAMILY_TRUMPET),
    Instrument("french-horn", "French Horn", CLEF_TREBLE, -7, FAMILY_HORN),
    Instrument("mellophone", "Mellophone", CLEF_TREBLE, -7, FAMILY_HORN),
    Instrument("alto-horn", "Alto / Tenor Horn", CLEF_TREBLE, -9, FAMILY_HORN),
    # Treble-clef baritone/euphonium read as a B♭ tenor instrument: a major
    # ninth (an octave plus a major second) below the written note.
    Instrument(
        "baritone-treble", "Baritone Horn (treble clef)", CLEF_TREBLE, -14, FAMILY_LOW_BRASS
    ),
    Instrument("euphonium-treble", "Euphonium (treble clef)", CLEF_TREBLE, -14, FAMILY_LOW_BRASS),
    # Bass-clef low brass reads concert pitch — written is what sounds.
    Instrument("trombone", "Trombone", CLEF_BASS, 0, FAMILY_LOW_BRASS),
    Instrument("bass-trombone", "Bass Trombone", CLEF_BASS, 0, FAMILY_LOW_BRASS),
    Instrument("tuba", "Tuba", CLEF_BASS, 0, FAMILY_LOW_BRASS),
)

# For the model field's ``choices``. Kept derived so the tuple above stays the
# single place an instrument is added or renamed.
INSTRUMENT_CHOICES: list[tuple[str, str]] = [(i.slug, i.label) for i in INSTRUMENTS]

_BY_SLUG: dict[str, Instrument] = {i.slug: i for i in INSTRUMENTS}


def get_instrument(slug: str) -> Instrument | None:
    """The instrument with this slug, or ``None`` if it isn't one we offer."""
    return _BY_SLUG.get(slug)

File: apps/mobile/src/data/instruments.ts
/**
 * The brass instruments a player can choose during onboarding.
 *
 * Mirrors the backend's `backend/users/instruments.py` (which owns the slugs the
 * API validates against) — change both together. It is duplicated rather than
 * fetched so the picker renders instantly and offline, and because the notation
 * layer will need this metadata on-device to draw transposed studies.
 *
 * The duplication is checked, not trusted: a backend test
 * (`InstrumentCatalogTests.test_typescript_mirror_matches_this_module`) parses
 * the `INSTRUMENTS` array below and fails if any slug, label, clef, offset, or
 * family disagrees with the Python module. Keep the array a flat list of object
 * literals so that reader keeps working.
 *
 * Transposition convention matches `StudyContent.transposition_semitones`:
 *
 *     sounding pitch = written pitch + soundingOffsetSemitones
 *
 * Onboarding reads only `slug` / `label` / `family`; `clef` and
 * `soundingOffsetSemitones` are groundwork for per-instrument transposition.
 */

export type Clef = 'treble' | 'bass';
export type InstrumentFamily = 'trumpet' | 'horn' | 'low-brass';

export interface Instrument {
  slug: string;
  label: string;
  clef: Clef;
  /** Semitones from written to sounding pitch; negative sounds below the page. */
  soundingOffsetSemitones: number;
  family: InstrumentFamily;
}

/** Picker order: the trumpet family Clarke wrote for, then horns, then low brass. */
export const INSTRUMENTS: Instrument[] = [
  { slug: 'trumpet', label: 'Trumpet', clef: 'treble', soundingOffsetSemitones: -2, family: 'trumpet' },
  { slug: 'cornet', label: 'Cornet', clef: 'treble', soundingOffsetSemitones: -2, family: 'trumpet' },
  { slug: 'flugelhorn', label: 'Flugelhorn', clef: 'treble', soundingOffsetSemitones: -2, family: 'trumpet' },
  // B♭ piccolo trumpet sounds an octave above the B♭ trumpet: -2 + 12.
  { slug: 'piccolo-trumpet', label: 'Piccolo Trumpet', clef: 'treble', soundingOffsetSemitones: 10, family: 'trumpet' },
  { slug: 'french-horn', label: 'French Horn', clef: 'treble', soundingOffsetSemitones: -7, family: 'horn' },
  { slug: 'mellophone', label: 'Mellophone', clef: 'treble', soundingOffsetSemitones: -7, family: 'horn' },
  { slug: 'alto-horn', label: 'Alto / Tenor Horn', clef: 'treble', soundingOffsetSemitones: -9, family: 'horn' },
  // Treble-clef baritone/euphonium read as a B♭ tenor instrument: a major ninth below written.
  { slug: 'baritone-treble', label: 'Baritone Horn (treble clef)', clef: 'treble', soundingOffsetSemitones: -14, family: 'low-brass' },
  { slug: 'euphonium-treble', label: 'Euphonium (treble clef)', clef: 'treble', soundingOffsetSemitones: -14, family: 'low-brass' },
  // Bass-clef low brass reads concert pitch — written is what sounds.
  { slug: 'trombone', label: 'Trombone', clef: 'bass', soundingOffsetSemitones: 0, family: 'low-brass' },
  { slug: 'bass-trombone', label: 'Bass Trombone', clef: 'bass', soundingOffsetSemitones: 0, family: 'low-brass' },
  { slug: 'tuba', label: 'Tuba', clef: 'bass', soundingOffsetSemitones: 0, family: 'low-brass' },
];

/** Section headings for the picker, in the order the families appear above. */
export const INSTRUMENT_FAMILY_LABELS: Record<InstrumentFamily, string> = {
  trumpet: 'Trumpet family',
  horn: 'Horns',
  'low-brass': 'Low brass',
};

const BY_SLUG = new Map(INSTRUMENTS.map((i) => [i.slug, i]));

/** The instrument with this slug, or undefined if it isn't one we offer. */
export function getInstrument(slug: string | null | undefined): Instrument | undefined {
  return slug ? BY_SLUG.get(slug) : undefined;
}

/** Display label for a stored slug, falling back to an em dash when unset. */
export function instrumentLabel(slug: string | null | undefined): string {
  return getInstrument(slug)?.label ?? '—';
}
